### ISSUE TITLE
Multiple queries in one transaction

### ISSUE THREAD

**Issue by chowey** [brianc/node-postgres#84]
Is there a way to get multiple queries to execute within one transaction?

Basically, I want to get the results from one query, and from its result branch to different queries, all within one transaction.  That way I can roll back if I need to, or if some error occurs.

Is this already supported for something like:

``` javascript
client.query('BEGIN', function (err, qry) {
  client.query('SELECT id FROM my_table WHERE order=$1', [order],
    function (err, qry) {
      if (qry.rowCount === 0) {
        client.query('INSERT INTO my_table(order) VALUES ($1)', [order],
          function (err, qry) {
            client.query('COMMIT');
          });
      } else {
        client.query('UPDATE my_table SET count=count+1 WHERE id=$1', [qry.rows[0].id],
          function (err, qry) {
            client.query('COMMIT');
          });
      }
    });
});
```

Per my experiments, it doesn't seem to execute in one transaction but correct me if I'm wrong.

**Comment by rishcd** [comment]
Dear Brianc,
I am trying to insert multiple rows in a table with serial PK. This inserted Pk has to be inserted in another table without affecting the relation logic. Please help me. below is the code iam working on
		pg.connect(connString1, function(err, client, done) {
    	if(err) {console.log('error fetching client from pool', err) }
    	 var query="insert into student_master VALUES(default,null,$1,$2,$3,$4,now(),$5,$6) returning id;";
		client.query(query,
		[data['rollNo'],
		data['name'],
		req.body.classId,
		data['regNo'],
		req.session.userId,
		data['vehicleNo']],
		function(err, result) {
    	if(err) {console.log('error inserting student ', err) }
   		var studentId=result.rows[0].id;
		console.log("student id obtained ::"+studentId)
			if((guardianList.length>0)&&studentId)
			{   
			function insertData(guardianList,callback)
			{
				var query2="insert into userinfo values(null,$1,default,$2,$3,null,$4,null,null,null) returning id"	;
				client.query(query2, [
				guardianList.guardianName,
				guardianList.email,
				passwordHash.generate(guardianList.pwd),
				req.session.userId
				],  function(err,result2) 
				{
					if(err) {console.log('error inserting userinfo ', err) }
					userId=result2.rows[0].id;
					console.log("userId  obtained in iteration is ::"+userId)
					var query3="insert into student_guardian values($1,$2,null,$3,now())"
					client.query(query3,[studentId,userId,req.session.userId],function(err,result3)
					{
						if(err) { return console.error('error running query', err)}
						console.log("user Id obtaiined for query 4 is "+userId)	
						var query4="insert into phone_user values(default,$1,$2)" 
						client.query(query4,[guardianList.guardianNo,userId],function(err,result3)
						{	//done();
						if(err) { return console.error('error running query phone update', err)}	
						})
					})
					callback(err);
				})
			}
			async.each(guardianList,insertData,function(err) {
			// Release the client to the pg module
			done();
			if (err) {console.log('error inserting ........... ', err) }
				res.send({"success":"success"})		
			})
	
			}
			else
			{
				done();
			}
				
   		    })
		});

**Comment by rishcd** [comment]
sometimes multiple insert happens to phone_user table(as expected) but with the same userId. I dont know how to tackle this. Please help

**Comment by rishcd** [comment]
let me rephrase the case once again for clarity
4 tables involved 
1. student_master
2. userinfo
3. student_guardian
4. phone_user

insert to student table will be only one
insert to the remaining tables are based on user input ie multiple records may come(multiple guardian for a student)
so for the multiple record, insert should first happen to userinfo and that will give the newly inserted serial PK. This PK should be inserted to student_guardian table and then inserted to phone_user (with phone no: from user input)

Hope the case is clear. 
Thanking you in anticipation

**Comment by rishcd** [comment]
userId  obtained in iteration is ::81
userId  obtained in iteration is ::82
user Id obtaiined for query 4 is 82
user Id obtaiined for query 4 is 82

copied from server console

**Comment by rishcd** [comment]
it should have been  81 82 81 82

**Comment by rishcd** [comment]
Dear Briac, sorry for so many posts

Now i solved the issue by changing the query to var query3="insert into student_guardian values($1,$2,null,$3,now()) returning guardian_id" and using this returned guardian_id instead of user_id(both same)
like below
var guardianId=result3.rows[0].guardian_id;
console.log("user Id obtaiined for query 4 is "+guardianId)	
var query4="insert into phone_user values(default,$1,$2)" 
client.query(query4,[guardianList.guardianNo,guardianId],function(err,result3)...

can you explain whats happening.. this is very twisty!!!!!!!!!!!!!!!!!!!!!!!! I really wonder how you people manage to write such difficult things ... hats off..   
I wish to get an explanation ( if u get time)..
